### PR TITLE
#ENG-93867 Document and lint static filter on Freedom UI detail datasource

### PR DESCRIPTION
**Pull request by b-horodyskyi**
## Jira
[ENG-93867](https://creatio.atlassian.net/browse/ENG-93867) — *Freedom UI: no clear/working way (and no agent guidance) to add a STATIC filter to a detail / related-list datasource.*

## Problem
A static business filter on a Freedom UI detail (`crt.DataGrid` / `crt.EntityDataSource`) had no documented, runtime-honored mechanism, and the clio MCP guidance didn't cover it. Agents fell back to `config.filters` on the viewElement datasource — **silently ignored** at runtime (the source reads only `entitySchemaName` + `attributes`), so the detail showed **unfiltered** data while `update-page` reported success. Contrast: `crt.IndicatorWidget` works because its filter lives inside `config.data.providing.filters`.

## Change
- **`related-list` guidance (#1):** documents the canonical static-filter mechanism — the collection attribute's `modelConfig.filterAttributes` + a sibling `<CollectionAttr>_PredefinedFilter` attribute holding `{ value: <ESQ group> }`, coexisting with `dependencies` (AND) and working standalone. Adds the `config.filters`-on-datasource **anti-pattern** to Common mistakes. Verified against runtime source (`entity-data-source.ts`, `base-view-model.ts`, `model-parameters-builder`) and real Freedom Designer output.
- **`PageBodyAstLinter` (#3):** new **Warning**-severity rule `entity-data-source-static-filters` flagging a `crt.EntityDataSource` whose `config` carries a `filters` key. Additive — never blocks the write; surfaces via `update-page` / `sync-pages` / `validate-page`.

## Tests
- 7 unit tests (`PageBodyAstLinterTests`): fires, no-filters, IndicatorWidget, `_PredefinedFilter`, type-gate isolation, multi-datasource targeting, finding anchor.
- 1 MCP E2E (`PageValidateToolE2ETests`): warning surfaces through the real MCP transport.
- Validated: `dotnet test --filter "FullyQualifiedName~PageBodyAstLinter"` (26/26, net8+net10); guidance tests 88/88; config.filters E2E green.
- End-to-end runtime check: agent authored a static-filtered detail via the guidance, used `_PredefinedFilter` (not `config.filters`), and the grid rendered correctly filtered in the browser.

## MCP / compatibility
MCP reviewed: guidance + lint rule aligned; no tool contract/args/destructive flags changed. ClioRing compatibility reviewed — no Ring-consumed contract changed (additive Warning only).

## Related
Component-doc counterpart (#2): creatio-ui `data-grid.component.md` (separate PR on `creatio.ghe.com/engineering/creatio-ui`).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

[ENG-93867]: https://creatio.atlassian.net/browse/ENG-93867?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by b-horodyskyi** [comment]
Component-doc counterpart (#2, creatio-ui `data-grid.component.md`): https://creatio.ghe.com/engineering/creatio-ui/pull/2714

File: clio/Command/McpServer/Resources/RelatedListGuidanceResource.cs [hunk 1]
```diff
@@ -193,11 +193,66 @@ dependency. Confirm the relationship with `get-app-info` / `get-entity-schema-pr
...
+		         the user changes (e.g. a `crt.SearchFilter` the user types into) AND (b) a STATIC business filter
+		         (see next section) — a fixed condition applied to the list. It is NEVER the record-scope mechanism:
+		         the designer emits `filterAttributes` ALONGSIDE `dependencies`, never instead of it. Use
+		         `dependencies` for "records that belong to the open record"; use `filterAttributes` for a static
+		         business filter or a user-driven filter ON TOP of that scope.
+
+		       Adding a STATIC business filter (Emails vs Activities, status/type-restricted lists)
+		       Some lists need a fixed condition — e.g. an "Emails" detail = `Activity` where `Type = Email`, an
+		       "Activities" detail = `Activity` where `Type != Email`, a "Current vacancies" list =
+		       `InternalRequest` where `Status IN (...)`, or a plain grid of `Contact` where `Account = <fixed>`.
```
[the remark this answers is not in the document]
Comment (b-horodyskyi) [reply clio/Command/McpServer/Resources/RelatedListGuidanceResource.cs:206]: Fixed in b22561a0. Updated the `related-list` catalog description + `[Description]` to name the `<CollectionAttr>_PredefinedFilter` / `filterAttributes` static-filter mechanism (and the `config.filters` anti-pattern), not just record scoping; added a Pages routing row → `related-list` for "add a detail / filter a list by the open record / apply a STATIC business filter"; and disambiguated the `business-rule-filters` row (that covers lookup/entity filters — a Freedom UI list's static filter is a page concern → `related-list`). Good catch, thanks.

File: clio/Command/McpServer/Tools/PageBodyAstLinter.cs [hunk 2]
```diff
@@ -270,6 +272,50 @@ private static void CheckConvertersDirectKeys(ObjectExpression convertersObj, Li
...
+		Property configProp = null;
+		bool isEntityDataSource = false;
+		foreach (Node element in obj.Properties) {
+			if (!TryGetInitProperty(element, out Property prop, out string key)) {
+				continue;
+			}
+			if (key == "type" && prop.Value is Literal { Value: "crt.EntityDataSource" }) {
+				isEntityDataSource = true;
+			} else if (key == "config" && prop.Value is ObjectExpression) {
+				configProp = prop;
```
[the remark this answers is not in the document]
Comment (b-horodyskyi) [reply clio/Command/McpServer/Tools/PageBodyAstLinter.cs]: Fixed in b22561a0. Re-keyed the rule off the EntityDataSource **config signature** (a `filters` key co-located with `entitySchemaName`) instead of requiring the inline `type` wrapper, so it now also catches a datasource whose descriptor and config are split across separate merges. Zero new false-positive risk — `entitySchemaName` is unique to an EntityDataSource config (`crt.IndicatorWidget` providing uses `schemaName`). Residual gap documented in the rule comment: a `filters`-only narrow merge into a `[...,"config"]` path with no co-located `entitySchemaName` still needs diff-path semantics — out of scope for this advisory Warning. Added a split-descriptor test that now fires.

**Review by Alexandr-Kravchuk — APPROVED** [review]

**Review by m-dymytrova — COMMENTED** [review]
## Review summary

**Decision: approve (minor issues)**

### Scoping note
The PR's base branch (`uc-337`) is stale relative to `master`, so the raw diff GitHub reports (50 files, +2877/-154) also contains two unrelated feature bundles (ENG-91254 SysSetting business-rule operand, ENG-93783 mobile support for requests) that were already reviewed and merged separately via PR #941 and PR #943, and only appear here through base-branch drift. This review is scoped to the actual ENG-93867-owned commits (`384667ef1`, `b22561a0`, `65d3f9d0` — 7 files, +339/-8): `PageBodyAstLinter.cs`, `RelatedListGuidanceResource.cs`, `RoutingGuidanceResource.cs`, `GuidanceCatalog.cs`, `PageBodyAstLinterTests.cs`, `PageValidateToolE2ETests.cs`, `.codex/workspace-diary.md`.

### On the prior approval and the bot findings
Both `chatgpt-codex-connector`'s P2 findings (routing/catalog description drift; the type-gated rule missing split-descriptor diffs) are genuinely fixed in `b22561a0`, and I confirmed this against the current code, not just the author's replies. The post-approval commit `65d3f9d0` (wording-accuracy fix responding to a Copilot comment on the sibling creatio-ui PR) is a pure comment/message reword with no logic change — I verified no test asserts the old wording, so it's safe.

### Findings (all Minor — no Blocker/Major)

1. **The split-descriptor fix still misses the single most common real-world trigger** (`PageBodyAstLinter.cs`, `CheckEntityDataSourceStaticFilters`). The rule now keys off `filters` + `entitySchemaName` co-located on the same object, closing the exact shape Codex quoted. But this repo's own platform guidance (`PageModificationOverviewGuidanceResource.cs:79-85`) mandates sending a **minimal delta body** when editing an *existing* page ("do NOT resend the full raw.body... compose a MINIMAL body that contains only the NEW operations"). An agent retrofitting a filter onto an *existing* `EmailDS` data source — exactly the "Emails vs Activities on an existing detail" scenario this ticket is about — will therefore send `{"operation":"merge","path":["dataSources","EmailDS","config"],"values":{"filters": {...}}}` with **no `entitySchemaName` in the submitted body at all** (it was set in an earlier call and correctly not resent). The rule doesn't fire on this. Fix direction: also key off the enclosing merge/replace operation's `path` array (`["dataSources", "<name>", "config"]`) — `filters` is never a valid `EntityDataSource` config key regardless of what else is in that operation's `values`, so this needs no diff-semantics state tracking, just reading the `path` array already present in the same parsed AST. Non-blocking (Warning-only, no regression vs. pre-PR — this shape was never caught before either), but worth a fast follow-up since the fix is cheap and closes the gap for the primary use case.
2. **The rule is unscoped beyond that signature** — it fires on *any* `ObjectExpression` anywhere in the page body (handlers, converters, arbitrary literals) that happens to pair `filters` + `entitySchemaName`, not just inside `modelConfig.dataSources`. I checked the current Creatio component/request catalog (`IndicatorWidget`, `ChartWidget`, the `PageSchemaHandlersGuidanceResource` request table) and found no shape that would collide today, so practical false-positive risk is low — but the same path-based fix in (1) would also scope this correctly as a side benefit.
3. Cosmetic: `65d3f9d0` reworded the production `Message` string and rule comment but left the matching unit test's `because:` text with the old "reads only entitySchemaName + attributes" phrasing (`PageBodyAstLinterTests.cs`, `Lint_ShouldEmitWarning_WhenEntityDataSourceHasConfigFilters`). No assertion depends on it — pure drift.
4. Optional/negligible: `CheckSchemaSectionShapes` and `CheckEntityDataSourceStaticFilters` each independently enumerate the same `ObjectExpression.Properties`; folding the new key checks into the existing loop (mirroring the `converters` branch already there) would save a redundant pass. No measurable perf impact given typical page-body sizes.

### Other dimensions
- **Security**: clean — no injection surface (pure AST inspection, no new I/O/auth path), and the `_PredefinedFilter` mechanism is not an access-control boundary change (row-level security is enforced server-side regardless of which filter mechanism the page uses). SonarQube Cloud's own gate on this PR shows 0 new issues / 0 new hotspots, consistent with this.
- **Testing**: verified all 7 claimed unit tests and the 1 E2E test are real and assert exactly what the PR description claims, with correct AAA/`because:`/`[Description]` compliance. The E2E test correctly proves the Warning-not-Error contract (`valid: true` with the rule id in `Warnings`).
- **Performance**: no complexity change — still O(N) over the AST, no new I/O or recursion.

### Requirement coverage
- Canonical `_PredefinedFilter` static-filter mechanism documented: **covered**.
- `config.filters` anti-pattern flagged via lint: **covered for inline + same-object split-descriptor shapes; gap for the minimal-delta/incremental-merge shape** (Finding 1).
- Additive/non-blocking (Warning severity): **covered**, verified via E2E.
- Routing/catalog discoverability: **covered**.
- Test coverage: **covered**.

### Confidence limits
- No live Jira/Confluence access in this session, so requirement mapping is derived from the PR description text, not the formal ENG-93867 acceptance criteria.
- Review scoped explicitly away from the ENG-91254/ENG-93783 content bundled in via base-branch drift, per the scoping note above — that content was not re-reviewed here.

**Review by m-dymytrova — APPROVED** [review]

**Review by tetiana-moshon — COMMENTED** [review]
## Agent-assisted PR review (Step 3) — decision: approve (minor issues)

Five-lens review (code-quality, security, performance, testing, bug-risk) run in parallel against the full diff (base `uc-337`, head `b288e8b3`). No Blocker or high-confidence Major finding. The primary in-scope change (ENG-93867) is high quality; all findings below are Minor / advisory, so no inline comments were posted.

### Scope / confidence note
The diff bundles **three independent concerns** because base `uc-337` is behind master: **ENG-93867** (the stated purpose — static-filter guidance + lint rule, genuinely new here) plus **ENG-91254** (business-rule `SysSetting` operand) and **ENG-93783** (mobile `get-request-info`), both of which already merged to master via #941 / #943 and appear only as a forward-port. Findings against the latter two are therefore against already-reviewed code, not net-new risk introduced by this PR. `mergeable: false` is a base-out-of-date artifact, not a review blocker.

### Findings (all Minor)

1. **Residual lint gap is documented in prose but not pinned by a test** — `clio.tests/Command/McpServer/PageBodyAstLinterTests.cs`. `CheckEntityDataSourceStaticFilters` keys off co-located `filters` + `entitySchemaName`; the one intentionally out-of-scope shape (a `filters`-only narrow merge with no sibling `entitySchemaName`) is called out in the source comment but has no negative test asserting it produces no finding, while every other boundary does. Add a one-line negative characterization test so the stated boundary is executable and a future widening can't silently regress it.

2. **Signature-gating slightly narrows detection vs the prior type-gate** — `clio/Command/McpServer/Tools/PageBodyAstLinter.cs` (`CheckEntityDataSourceStaticFilters`). An inline full `crt.EntityDataSource` descriptor that omits `entitySchemaName` in its `config` is no longer flagged (the previous type-gated rule caught it). Practical exposure is small — a real EntityDataSource almost always carries `entitySchemaName`, and the rule is a non-blocking Warning. Optional: OR the signature gate with an enclosing `type: "crt.EntityDataSource"` check to restore that shape without reintroducing the split-merge miss. The IndicatorWidget non-fire (`schemaName` vs `entitySchemaName`) and the `filters`-key line/column anchoring were both verified correct.

3. **`ValidateComparison` family relaxation widens all operand pairs, not only SysSetting** — `clio/Command/BusinessRules/BusinessRuleValidator.cs` (`ValidateComparison`). The new text-family / numeric-family `sameType` check applies to every condition, so attribute-vs-attribute pairs previously rejected (e.g. `ShortText` vs `LongText`, `Integer` vs `Money`) are now accepted. It is a widening (cannot reject a previously-valid rule), DateTime/Lookup stay exact, and the "verified on a live environment" claim covers the motivating SysSetting case. This is ENG-91254 code already merged via #941; recommend a follow-up test asserting attribute-vs-attribute cross-subtype acceptance is intended.

4. **No cross-rule sys-setting resolution cache in batch create/update** — `clio/Command/BusinessRules/{EntityBusinessRuleService,PageBusinessRuleService}.cs`. `Resolve(rule)` dedups setting codes within a single rule only, so a batch of N rules gating on the same setting issues N identical DataService round-trips (the batch already caches the static-filter context this way). Bounded — batches are small and user-authored, and rules with no SysSetting operand make zero round-trips. Optional: hoist a batch-scoped descriptor cache.

5. **Converter treats a missing SysSetting map entry two ways** — `clio/Command/BusinessRules/Converters/SimpleToFullBusinessRuleConverter.cs`. `ResolveOperandTypeContext` returns `null` on a map miss while `BuildOperandExpression` throws on the same miss. Benign today (the validator rejects unresolved settings before conversion), but a latent inconsistency; make both throw the same invariant violation.

### Security
No credible vulnerabilities. Notably well-defended: `GetSysSettingTypeByCode` reads only setting **type** metadata (never the value) and hard-rejects `SecureText`/`Binary` operands, keeping secret-bearing settings out of client-side rule conditions; the code lookup is ESQ-parameterized (no injection); error paths preserve `SensitiveErrorTextRedactor.Redact`; the docs-path traversal guards are unchanged. The `System.Security.Cryptography.Xml` posture is safe (past the CVE-2026-33116 fix line, with NU1903 promoted to error).

### Requirement coverage (ENG-93867)
- **R1** — document the canonical static-filter mechanism: **covered** (RelatedListGuidanceResource + the `crt.DataGrid` component doc as SHAPE owner).
- **R2** — expose it in clio MCP guidance + routing / get-component-info: **covered** (routing map, GuidanceCatalog, resource `[Description]` all updated to the `_PredefinedFilter`/`filterAttributes` mechanism and the `config.filters` anti-pattern; component-doc counterpart in separate creatio-ui PR).
- **R3** — validate/reject the invalid `config.filters` shape at update-page: **covered** as a non-blocking Warning (`entity-data-source-static-filters`), a defensible reading of "consider … (currently accepted silently)"; surfaces through update-page / sync-pages / validate-page with unit + real-transport E2E coverage.

Tests: strong across the board — explicit AAA, `[Description]` on every method, `because:` on assertions, cross-OS-safe fixtures, positive/negative pairing, and 4 new MCP E2E tests proving behavior survives the real transport. MCP maintenance and ClioRing-compatibility policies are addressed in the PR description (additive Warning; no tool contract change).

**Review by tetiana-moshon — APPROVED** [review]
